modules/core/src/main/java/org/apache/ignite/internal/tostring/IgniteLogThrottle.java

```diff
@@ -0,0 +1,532 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.tostring;
+
+import static org.apache.ignite.internal.util.ArrayUtils.OBJECT_EMPTY_ARRAY;
+import static org.apache.ignite.internal.util.ArrayUtils.nullOrEmpty;
+
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import org.apache.ignite.internal.lang.IgniteStringFormatter;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.util.FastTimestamps;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Log throttle.
+ *
+ * <p>Errors are logged only if they were not logged for the last {@link #THROTTLE_TIMEOUT_MILLIS} milliseconds. Note that not only error
+ * messages are checked for duplicates, but also exception classes.</p>
+ */
+public class IgniteLogThrottle {
+    /** Throttle timeout in milliseconds (value is 5 min). */
+    private static final long THROTTLE_TIMEOUT_MILLIS = TimeUnit.MINUTES.toMillis(5);
+
+    /** Log messages. */
+    private static final ConcurrentMap<LogThrottleKey, Long> MESSAGES_MAP = new ConcurrentHashMap<>(128, 0.75f);
+
+    /**
+     * Logs error if needed.
+     *
+     * @param log Logger.
+     * @param msg Message.
+     */
+    public static void error(IgniteLogger log, String msg) {
```

Comment:
Instead of using static methods, could we use inheritance and have something like `IgniteThrottledLogger`?

Reply:
Yes, I think it will look more natural.